### DISCUSSION TITLE
Upload & download data to&from ThirdParty Services like LibreView

### DISCUSSION THREAD

**Discussion by yoshimo** [Navid200/xDrip#96] [Ideas]
The Freestyle Libre2 App uses an account at Abbott to upload measurements and provide data for Libreview (doctors) and LibreLink for friends and family members to get access to alarms and measurements too.
Instead of using a patched app it would be nice if we could teach xdrip+ the necessary methods to do the upload directly.

**Comment by simonpickering** [comment]
Yes it would indeed be nice - there are also other sites that the doctors/consultants use, such as Diasend/Glooko. 

There are a couple of options re the Abbott site and app - eavesdrop on the comms between the app and the site, or see if Abbott would be willing to publish an API. 

I got in touch with Diasend in April this year to ask about an API and was fobbed off...

There may be existing code lurking on github that could be put to good use, even just establishing a list of sites/services to which data might need to be exported for doctor/consultant use would be a useful thing to do.

**Reply by yoshimo** [reply]
I honestly don't think we will see support for third parties from manufacturers anytime soon. Feature requests and technical issues are often answered with the same standard text. We are on our own, building software, analyzing the hardware interfaces and building helper tools.

Which other services and applications do we need to cover?

**Comment by simonpickering** [comment]
Off-hand I don't know, but I imagine we could ask and get a list running, with the goal of all upload endpoints being served as much as possible by the same core code.

**Comment by MarekM60** [comment]
Applications like LibreLink/...../xDrip/.... are medical devices too and need full clinical tests before distribution to end-users. From this point of view no one manufacturer (Abbott, Dexcom,...) will agree to formal use (disclose interfaces) of their products by community developers.

**Reply by simonpickering** [reply]
Yes indeed, though there's hopefully no reason why data export couldn't be enabled, which is a first (and perhaps only!) step. Hopefully the consultants/doctors will be willing to accept something like Tidepool as a site for review purposes.

**Comment by yoshimo** [comment]
Now that i think of it, instead of just limiting ourselfes to "upload" and one specific vendor, it would make also sense to use these services as an additional data source and pull data like the libre link app does in the freestyle universe.

**Comment by psonnera** [comment]
Best way to get xDrip+ taken down for DMCA IMHO.

**Reply by yoshimo** [reply]
based on which arguments? 
There are exceptions for making things compatible and fix bugs and add features.
I also doubt that you can claim copyright on a simple json/rest/xml service.

**Reply by simonpickering** [reply]
I'd have thought that scraping/downloading data would be fine, I think reverse engineering upload may cause more issues and might be better in a separately installed plugin, if it can indeed be made to work. But then I look to Tidepool, which works, and would suggest that to my consultant.

Something to be discussed once the capability exists I'd say.

**Comment by yoshimo** [comment]
Research was done with an implementation by
https://github.com/timoschlueter/nightscout-librelink-up
Which this could be based on

**Reply by Navid200** [reply]
You can continue to use these discussions if you wish.
But, you will have a larger audience if you switch to the discussions in the main repository:
https://github.com/NightscoutFoundation/xDrip/discussions

So, I strongly suggest that you switch.
These discussions will eventually be closed.
